### DISCUSSION TITLE
Question about takeaway from `drop_static<T: 'static>(t: T)`

### DISCUSSION THREAD

**Discussion by maxim-ge** [pretzelhammer/rust-blog#48] [Q&A] [answered]
Hi,

in the very interesting article [Common Rust Lifetime Misconceptions](https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md) you write:

> A program to help illustrate these concepts:
```rust
use rand;

fn drop_static<T: 'static>(t: T) {
    std::mem::drop(t);
}
```

What you think about [`drop_static<'nonstatic, T: 'nonstatic>(t: T)`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9d4b8ff08f8c66d14f83b87b29c43f82)?

It is compiled and run, too.

**Comment by pretzelhammer** [comment]
Can you please clarify your question? I'm not sure what you are asking. You can drop things which can live up to the end of the program, so you can certainly drop things which live shorter than that as well, so I'm not sure what the question here is.

**Reply by maxim-ge** [reply]
You claim that "T: 'static should be read as "T is bounded by a 'static lifetime". I think that it is just ignored by compiler in this case.

[Another example:](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=33caf64ebf061db0bce49c4d12ab8aae) 

```
fn print_static<T: 'static + std::fmt::Display>(t: T) {
    println!("{}", t);
}

fn main(){
    let s = String::from("Hello");
    print_static(s);
}
```

Or may be I'm wrong and `s` is bounded by static lifetime here?

**Reply by pretzelhammer** [reply]
`s` is bounded by `'static` where the bound is an upper bound. `s` _can_ be held indefinitely long, including up until the end of the program, but it also _can_ be dropped earlier. It doesn't have to live as long as `&'static T` but it could if it wanted to. Does that help explain?

**Reply by maxim-ge** [reply]
Could you please give an example of the call to print_static that does not compile (because of lifetime incompatibility)?

**Reply by WaffleLapkin** [reply, marked as answer]
```rust
fn print_static<T: 'static + std::fmt::Display>(t: T) {
    println!("{}", t);
}

fn main(){
    let s = String::from("Hello");
    print_static(&s); // `&s` can't live till the end of the program, since `s` is dropped
}
```
[[playground]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ad0c608fa05a70aff19971a190d560a4)

**Reply by maxim-ge** [reply]
@WaffleLapkin - thanks, now it is more clear
